eip/EIPS/eip-4337.md

```diff
@@ -360,16 +360,76 @@ Stated in simpler terms, we expect at least `1 / MIN_INCLUSION_RATE_DENOMINATOR`
+When `initCode` is specified, if either the `sender` address points to an existing contract, or (after calling the initCode) the `sender` address still does not exist,
+then the operation is aborted.
+The `initCode` MUST NOT be called directly from the entryPoint, but from another address.
+The contract created by this factory method should accept a call to `validateUserOp` to validate the UserOp's signature.
+For security reasons, it is important that the generated contract address will depend on the initial signature.
+This way, even if someone can create a wallet at that address, he can't set different credentials to control it.
+
+NOTE: In order for the wallet to determine the "counterfactual" address of the wallet (prior its creation), 
+it should make a static call to the `entryPoint.getSenderAddress()`
+
+### Entry point upgrading
+
+Wallets are encouraged to be DELEGATECALL forwarding contracts for gas efficiency and to allow wallet upgradability. The wallet code is expected to hard-code the entry point into their code for gas efficiency. If a new entry point is introduced, whether to add new functionality, improve gas efficiency, or fix a critical security bug, users can self-call to replace their wallet's code address with a new code address containing code that points to a new entry point. During an upgrade process, it's expected that two mempools will run in parallel.
+
+## RPC methods
+
+### * eth_sendUserOperation
+
+eth_sendUserOperation submits a User Operation object to the User Operation pool of the client. The client MUST validate the UserOperation, and return a result accordingly.
+
+The result `SHOULD` be set to the **requestId** if and only if the request passed simulation and was accepted in the client's User Operation pool. If the validation, simulation, or User Operation pool inclusion fails, `result` `SHOULD NOT` be returned. Rather, the client `SHOULD` return the failure reason.
+
+#### Parameters:
+1. **UserOperation** a full user-operation struct. All fields MUST be set as hex values. empty `bytes` block (e.g. empty `initCode`) MUST be set to `"0x"`
+2. **EntryPoint** the entrypoint address the request should be sent through. this MUST be one of the entry points returned by the `supportedEntryPoints` rpc call.
+
+#### Return value:
+* If the UserOperation is valid, the client MUST return the calculated **requestId** for it
+* in case of failure, MUST return an `error` result object, with `code` and `message`. The error code and message SHOULD be set as follows:
+  * **code: -32602** - invalid UserOperation struct/fields
+  * **code: -32500** - transaction rejected by entryPoint's simulateValidation.
+    * The `message` field SHOULD be set to the FailOp error message
+    * The `data` field SHOULD contain a `paymaster` value, if set by this FailedOp
+  * **code: -32501** - transaction rejected because paymaster is throttled/banned
```

Comment:
Missing a code for "paymaster is not sufficiently staked".  This one is for throttled/banned.  In the example below you use -33000 but it doesn't appear here.

Reply:
Ok.
(Codes are 32500, since I'm not sure if they deliberately selected numbers just close to type(int16).min ..
I didn't want to get below -32768..